### PR TITLE
Add nested multi-resolution volume render backend

### PR DESCRIPTION
**Pull request by ojessop01**
Implement efficient Sparse Multi-Scale Grid algorithm (Benitez-Llambay 2025)
for 3D volume rendering of SPH data. Particles are hierarchically assigned
to grid levels based on kernel smoothing length, then scattered onto
coarse-to-fine grids and upsampled back to target resolution. This bounds
per-particle work regardless of smoothing length, providing significant
speedups for simulations with wide kernel distributions.

Features:
- New nested grid backend accessible via backend="nested" parameter
- Adaptive hierarchy with configurable target voxels per kernel (ntarget)
  and number of levels (nlevels)
- Kernel arithmetic matches scatter.py implementation (Wendland-C2)
- Parallel support via Numba prange
- Trilinear upsampling for hierarchy collapse
- Efficient bounding-box tracking to skip empty regions

API additions:
- render_voxel_grid() and render_gas() now accept backend, ntarget, nlevels
- backends dict updated to include "nested" alongside "scatter"
- Added validation and warnings for backend-specific parameters

Tests:
- Added TestNestedVolumeRender class with serial/parallel matching tests
- Tests cover periodic boundary conditions and edge cases
- Comparison against single-resolution backend validates correctness

Documentation:
- Updated volume_render.rst with nested backend description and examples
- Added parameter documentation for ntarget and nlevels

Code quality:
- Passes ruff linter
- Passes numpydoc validation
- All existing tests continue to pass

### CODE REVIEW

**Comment by kyleaoman** [comment]
Thanks Owen, looks promising. Presumably this can be trivially translated to the 2D ("project") case? And maybe slice, too? Or do you think it's not worth it because that scales as N^2 instead of N^3?

**Comment by ojessop01** [comment]
Yes this could be trivially extended to 2d projection and slice and would scale much better than the current implementations. I just haven't had a chance to implement this and ensure all the formatting etc is correct and the documentation written.

**Comment by kyleaoman** [comment]
Great, and not a problem - marking this as draft for now if you plan to add that (if not feel free to mark ready for review).

**Comment by ojessop01** [comment]
i plan to add eventually but not sure when I can get around to it. I will submit this ready for review for now.

File: docs/source/visualisation/volume_render.rst [hunk 1]
```diff
@@ -380,6 +380,246 @@ Once you have this base image, you can always use your photo editor to tweak it
...
+
+   data = load("cosmo_volume_example.hdf5")
+
+   # Standard backend — identical results, familiar behaviour.
+   mass_grid_standard = render_gas(
+       data,
+       resolution=256,
+       project="masses",
+       parallel=True,
+       backend="scatter",
```
Comment (kyleaoman) [inline docs/source/visualisation/volume_render.rst:409]:
```suggestion
       backend="scatter",  # the default
```

File: docs/source/visualisation/volume_render.rst [hunk 2]
```diff
@@ -380,6 +380,246 @@ Once you have this base image, you can always use your photo editor to tweak it
...
+  and diffuse IGM.
+
+.. code-block:: python
+
+   from swiftsimio import load
+   from swiftsimio.visualisation.volume_render import render_gas
+
+   data = load("cosmo_volume_example.hdf5")
+
+   # Standard backend — identical results, familiar behaviour.
```
Comment (kyleaoman) [inline docs/source/visualisation/volume_render.rst:403]:
```suggestion
   # Standard backend
```

File: docs/source/visualisation/volume_render.rst [hunk 3]
```diff
@@ -380,6 +380,246 @@ Once you have this base image, you can always use your photo editor to tweak it
...
+simulations, or cosmological volumes that include both dense filaments and
+low-density IGM — contain particles whose kernels span tens or hundreds of
+voxels. These few large particles can completely dominate the total rendering
+time even though they represent only a tiny fraction of the particle count.
+
+The nested backend bounds the per-particle cost at
+:math:`\mathcal{O}(n_\mathrm{target}^3)` regardless of smoothing length by
+scattering large particles onto a coarser grid and trilinearly upsampling
+the result back to the finest grid afterwards. This follows the Sparse
+Multi-Scale Grid approach described in Benitez-Llambay (2025).
```
Comment (kyleaoman) [inline docs/source/visualisation/volume_render.rst:443]:
```suggestion
Multi-Scale Grid approach described in `Benítez-Llambay (2025)`_.

_Benítez-Llambay (2025): https://iopscience.iop.org/article/10.3847/2515-5172/addab2
```

File: docs/source/visualisation/volume_render.rst [hunk 4]
```diff
@@ -380,6 +380,246 @@ Once you have this base image, you can always use your photo editor to tweak it
...
+coarse cell is trilinearly upsampled and its contribution added to the next
+finer grid. The upsampling stencil is::
+
+   fine index 2i   draws 3/4 from coarse[i] and 1/4 from coarse[i-1]
+   fine index 2i+1 draws 3/4 from coarse[i] and 1/4 from coarse[i+1]
+
+applied independently in each of x, y, z, giving a separable
+trilinear interpolation. Only the bounding box of voxels that actually
+received mass is upsampled at each level, so the collapse cost scales with
+the number of particles rather than the grid volume.
```
Comment (kyleaoman) [inline docs/source/visualisation/volume_render.rst:499]:
```suggestion
received mass is upsampled at each level, so the collapse cost scales
approximately with the number of particles rather than the grid volume.
```

File: docs/source/visualisation/volume_render.rst [hunk 5]
```diff
@@ -417,11 +656,25 @@ raw numpy array (not :class:`~swiftsimio.objects.cosmo_array` or
...
 
-   volume_render_scatter = backends["scatter"]
-   # Using the variable names from above
-   out = volume_render_scatter(x=x, y=y, z=z, h=h, m=m, res=res)
+   # Nested multi-resolution scatter (serial).
+   out = backends["nested"](
+       x=x, y=y, z=z, h=h, m=m, res=res,
+       ntarget=6,
+       nlevels=4,
+   )
```
Comment (kyleaoman) [inline docs/source/visualisation/volume_render.rst:669]:
```suggestion
   out = backends["nested"](
       x=x,
       y=y,
       z=z,
       h=h,
       m=m,
       res=res,
       ntarget=6,
       nlevels=4,
   )
```

File: docs/source/visualisation/volume_render.rst [hunk 6]
```diff
@@ -417,11 +656,25 @@ raw numpy array (not :class:`~swiftsimio.objects.cosmo_array` or
...
+
+   # Parallel variants — use the same keyword arguments.
+   out = backends_parallel["scatter"](x=x, y=y, z=z, h=h, m=m, res=res)
+   out = backends_parallel["nested"](
+       x=x, y=y, z=z, h=h, m=m, res=res,
+       ntarget=6,
+       nlevels=4,
+   )
```
Comment (kyleaoman) [inline docs/source/visualisation/volume_render.rst:677]:
```suggestion
   out = backends_parallel["nested"](
       x=x,
       y=y,
       z=z,
       h=h,
       m=m,
       res=res,
       ntarget=6,
       nlevels=4,
   )
```

File: swiftsimio/visualisation/volume_render.py [hunk 7]
```diff
@@ -84,6 +88,24 @@ def render_voxel_grid(
...
+    backend : str, optional
+        The scatter backend to use. Available backends are ``"scatter"``
+        (default, standard single-resolution) and ``"nested"`` (nested
+        multi-resolution, faster for simulations with wide smoothing-length
+        distributions).
+
+    ntarget : int, optional
+        Only used with ``backend="nested"``. Target number of voxels per
+        kernel compact-support diameter at each level of the nested grid
+        hierarchy. Default is 6. A ``UserWarning`` is raised if this is set
```
Comment (kyleaoman) [inline swiftsimio/visualisation/volume_render.py:100]:
```suggestion
        hierarchy. Default is ``6``. A ``UserWarning`` is raised if this is set
```

File: swiftsimio/visualisation/volume_render.py [hunk 8]
```diff
@@ -84,6 +88,24 @@ def render_voxel_grid(
...
+        without ``backend="nested"``.
+
+    nlevels : int, optional
+        Only used with ``backend="nested"``. Number of coarsening levels in
+        the nested grid hierarchy. Default is 4. ``resolution`` must be
```
Comment (kyleaoman) [inline swiftsimio/visualisation/volume_render.py:105]:
```suggestion
        the nested grid hierarchy. Default is ``4``. ``resolution`` must be
```

File: swiftsimio/visualisation/volume_render.py [hunk 9]
```diff
@@ -197,6 +243,24 @@ def render_gas(
...
+    backend : str, optional
+        The scatter backend to use. Available backends are ``"scatter"``
+        (default, standard single-resolution) and ``"nested"`` (nested
+        multi-resolution, faster for simulations with wide smoothing-length
+        distributions).
+
+    ntarget : int, optional
+        Only used with ``backend="nested"``. Target number of voxels per
+        kernel compact-support diameter at each level of the nested grid
+        hierarchy. Default is 6. A ``UserWarning`` is raised if this is set
```
Comment (kyleaoman) [inline swiftsimio/visualisation/volume_render.py:255]:
```suggestion
        hierarchy. Default is ``6``. A ``UserWarning`` is raised if this is set
```

File: swiftsimio/visualisation/volume_render.py [hunk 10]
```diff
@@ -197,6 +243,24 @@ def render_gas(
...
+        without ``backend="nested"``.
+
+    nlevels : int, optional
+        Only used with ``backend="nested"``. Number of coarsening levels in
+        the nested grid hierarchy. Default is 4. ``resolution`` must be
```
Comment (kyleaoman) [inline swiftsimio/visualisation/volume_render.py:260]:
```suggestion
        the nested grid hierarchy. Default is ``4``. ``resolution`` must be
```

File: tests/test_visualisation.py [hunk 11]
```diff
@@ -1002,6 +1002,333 @@ def test_equivalent_regions(self, cosmological_volume_only_single_local):
...
+            m=np.array([1.0, 1.0, 1.0, 1.0]),
+            h=np.array([0.2, 0.2, 0.2, 0.000_002]),
+            res=64,
+            box_x=1.0,
+            box_y=1.0,
+            box_z=1.0,
+        )
+
+    def test_volume_parallel(self):
+        """Check that nested parallel render matches serial nested render."""
```
Comment (kyleaoman) [inline tests/test_visualisation.py:1024]:
```suggestion
        """Check that parallel nested render matches serial nested render."""
```

File: tests/test_visualisation.py [hunk 12]
```diff
@@ -1002,6 +1002,333 @@ def test_equivalent_regions(self, cosmological_volume_only_single_local):
         )
 
 
+class TestNestedVolumeRender:
```
Comment (kyleaoman) [inline tests/test_visualisation.py:1005]: Most of these tests look near-identical to the ones for the default scatter backend. Should refactor to parametrize the existing tests over the available backends. I think that the only exception is the comparison of the nested backend to the standard backend.

File: swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py [hunk 13]
```diff
@@ -0,0 +1,1225 @@
...
+3. Scatter each level's particles onto their coarse grid.
+
+4. Collapse the hierarchy coarse-to-fine: trilinearly upsample each coarse
+   grid and accumulate it into the next finer grid.
+
+5. Return the finest grid.
+
+The algorithm follows a novel Sparse Multi-Scale Grid algorithm
+(as described in Benitez-Llambay 2025) to place particles on a grid using
+an Adaptive Mesh Refinement (AMR) approach.
```
Comment (kyleaoman) [inline swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py:36]:
```suggestion
The algorithm follows a novel Sparse Multi-Scale Grid algorithm
(as described in Benitez-Llambay 2025, doi:10.3847/2515-5172/addab2)
to place particles on a grid using an Adaptive Mesh Refinement (AMR)
approach.
```

File: swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py [hunk 14]
```diff
@@ -0,0 +1,1225 @@
...
+    m: np.ndarray,
+    res: int,
+    ntarget: int,
+    nlevels: int,
+) -> tuple:
+    """
+    Assign a hierarchy level to every particle in a single pass.
+
+    Replaces ``ceil(log2(support_cells / ntarget))`` with a cheap doubling
+    loop, avoiding transcendental calls in the hot path.  Particles with
```
Comment (kyleaoman) [inline swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py:68]: "transcendental calls in the hot path"?? Translate to English?

File: swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py [hunk 15]
```diff
@@ -0,0 +1,1225 @@
...
+    x_pos : float64
+        Particle x coordinate in [0, 1].
+    y_pos : float64
+        Particle y coordinate in [0, 1].
+    z_pos : float64
+        Particle z coordinate in [0, 1].
+    mass_icv : float32
+        ``mass * level_res**3`` — used by the sub-pixel deposit path.
+    weighted_prefactor : float32
+        ``mass * (21/2π) * H**-3`` — kernel amplitude folded into one scalar.
```
Comment (kyleaoman) [inline swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py:170]:
```suggestion
        ``mass * (21 / 2 / np.pi) * h**-3`` — kernel amplitude folded into one scalar.
```

File: swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py [hunk 16]
```diff
@@ -0,0 +1,1225 @@
...
+            destination = coarse
+            level_offset = level_offsets[level] - finest_cells
+
+        level_res = level_resolutions[level]
+        level_plane_p = level_res * level_res
+        original_x = x[particle]
+        original_y = y[particle]
+        original_z = z[particle]
+
+        # Hoist all particle-invariant constants before the periodic image loop.
```
Comment (kyleaoman) [inline swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py:442]: Hoist?

File: swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py [hunk 17]
```diff
@@ -0,0 +1,1225 @@
...
+            if fine_y_stop - 1 > bounds_max[level - 1, 1]:
+                bounds_max[level - 1, 1] = fine_y_stop - 1
+            if fine_z_stop - 1 > bounds_max[level - 1, 2]:
+                bounds_max[level - 1, 2] = fine_z_stop - 1
+
+
+@njit(cache=True, fastmath=True, nogil=True)
+def _build_hierarchy_layout(res: int, nlevels: int) -> tuple:
+    """
+    Build flattened hierarchy metadata and per-level hot-loop constants.
```
Comment (kyleaoman) [inline swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py:700]: Hot-loop?

File: swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py [hunk 18]
```diff
@@ -0,0 +1,1225 @@
...
+
+    JIT-callable wrapper around the full assign → scatter → collapse pipeline,
+    designed to be invoked from ``prange`` inside the parallel kernel and as
+    the single-thread entry point from the public API.  Each call builds its
+    own private hierarchy and accumulates into a private finest grid.
+
+    Parameters
+    ----------
+    x, y, z : np.ndarray[float64]
+        Particle positions for this chunk, in [0, 1].
```
Comment (kyleaoman) [inline swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py:782]:
```suggestion
    x : np.ndarray[float64]
        Particle x-positions for this chunk, in [0, 1].
    y : np.ndarray[float64]
        Particle y-positions for this chunk, in [0, 1].
    z : np.ndarray[float64]
        Particle z-positions for this chunk, in [0, 1].
```

File: swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py [hunk 19]
```diff
@@ -0,0 +1,1225 @@
...
+    m : np.ndarray[float32]
+        Particle masses or weights for this chunk.
+    h : np.ndarray[float32]
+        Smoothing lengths for this chunk.
+    res : int
+        Finest-grid resolution (cells per axis).
+    box_x, box_y, box_z : float64
+        Periodic box sizes; 0.0 disables wrapping on that axis.
```
Comment (kyleaoman) [inline swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py:790]:
```suggestion
    box_x : float64
        Periodic box size in x-direction; 0.0 disables wrapping on that axis.
    box_y : float64
        Periodic box size in y-direction; 0.0 disables wrapping on that axis.
    box_z : float64
        Periodic box size in z-direction; 0.0 disables wrapping on that axis.
```

File: swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py [hunk 20]
```diff
@@ -0,0 +1,1225 @@
...
+    shared output after all chunks complete.
+
+    Particle-level parallelism (rather than voxel-level) is used because
+    different particles visit different subsets of voxels; voxel-level
+    splitting would require synchronisation on every write.
+
+    Parameters
+    ----------
+    x, y, z : np.ndarray[float64]
+        Contiguous particle positions in [0, 1].
```
Comment (kyleaoman) [inline swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py:885]:
```suggestion
    x : np.ndarray[float64]
        Contiguous particle x-positions in [0, 1].
    y : np.ndarray[float64]
        Contiguous particle y-positions in [0, 1].
    z : np.ndarray[float64]
        Contiguous particle z-positions in [0, 1].
```

File: swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py [hunk 21]
```diff
@@ -0,0 +1,1225 @@
...
+    m : np.ndarray[float32]
+        Contiguous particle masses or weights.
+    h : np.ndarray[float32]
+        Contiguous particle smoothing lengths.
+    res : int
+        Finest-grid resolution (cells per axis).
+    box_x, box_y, box_z : float64
+        Periodic box sizes; 0.0 disables wrapping on that axis.
```
Comment (kyleaoman) [inline swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py:893]:
```suggestion
    box_x : float64
        Periodic box size in x; 0.0 disables wrapping on that axis.
    box_y : float64
        Periodic box size in y; 0.0 disables wrapping on that axis.
    box_z : float64
        Periodic box size in z; 0.0 disables wrapping on that axis.
```

File: swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py [hunk 22]
```diff
@@ -0,0 +1,1225 @@
...
+            box_y,
+            box_z,
+            ntarget,
+            nlevels,
+        )
+    return output
+
+
+# ---------------------------------------------------------------------------
+# Public API
```
Comment (kyleaoman) [inline swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py:934]: The first method below is private?

File: swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py [hunk 23]
```diff
@@ -0,0 +1,1225 @@
...
+    Validate all inputs and return typed, contiguous arrays with scalar config.
+
+    Positions (x, y, z) are promoted to float64 to avoid rounding errors in
+    cell-index arithmetic at high resolution.  Masses and smoothing lengths
+    (m, h) are cast to float32 to match the Numba kernel signatures.
+
+    Parameters
+    ----------
+    x, y, z : array-like
+        Particle positions; must be 1-D with equal length.
```
Comment (kyleaoman) [inline swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py:958]:
```suggestion
    x : array-like
        Particle x-positions; must be 1-D with equal length.
    y : array-like
        Particle y-positions; must be 1-D with equal length.
    z : array-like
        Particle z-positions; must be 1-D with equal length.
```

File: swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py [hunk 24]
```diff
@@ -0,0 +1,1225 @@
...
+        Maximum hierarchy depth.
+
+    Returns
+    -------
+    x, y, z : np.ndarray[float64]
+        Contiguous position arrays.
+    m, h : np.ndarray[float32]
+        Contiguous mass and smoothing-length arrays.
+    res, ntarget, nlevels : int
+        Validated integer scalars.
```
Comment (kyleaoman) [inline swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py:977]:
```suggestion
    x : np.ndarray[float64]
        Contiguous x-position array.
    y : np.ndarray[float64]
        Contiguous y-position array.
    z : np.ndarray[float64]
        Contiguous z-position array.
    m : np.ndarray[float32]
        Contiguous mass array.
    h : np.ndarray[float32]
        Contiguous smoothing-length array.
    res : int
        Finest-grid resolution (cells per axis).
    ntarget : int
        Target cells per kernel side.
    nlevels : int
        Maximum hierarchy depth.
```

File: swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py [hunk 25]
```diff
@@ -0,0 +1,1225 @@
...
+        Box size in y, in the same rescaled length units as x, y and z.
+        Used for periodic wrapping. Set to 0.0 to disable wrapping in y.
+
+    box_z : np.float64
+        Box size in z, in the same rescaled length units as x, y and z.
+        Used for periodic wrapping. Set to 0.0 to disable wrapping in z.
+
+    ntarget : int
+        Target number of voxels spanned by each kernel at its assigned level.
+        Controls the accuracy/speed trade-off. Default is 6.
```
Comment (kyleaoman) [inline swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py:1111]:
```suggestion
        Controls the accuracy/speed trade-off. Default is ``6``.
```

File: swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py [hunk 26]
```diff
@@ -0,0 +1,1225 @@
...
+
+    nlevels : int
+        Maximum number of coarsening levels in the hierarchy. Default is 4.
```
Comment (kyleaoman) [inline swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py:1114]:
```suggestion
        Maximum number of coarsening levels in the hierarchy. Default is ``4``.
```

File: swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py [hunk 27]
```diff
@@ -0,0 +1,1225 @@
...
+
+    Returns
+    -------
+    np.ndarray[np.float32, np.float32, np.float32]
```
Comment (kyleaoman) [inline swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py:1118]: Is content of brackets correct? Thought it was type then shape?

File: swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py [hunk 28]
```diff
@@ -0,0 +1,1225 @@
...
+        Box size in y, in the same rescaled length units as x, y and z.
+        Used for periodic wrapping. Set to 0.0 to disable wrapping in y.
+
+    box_z : np.float64
+        Box size in z, in the same rescaled length units as x, y and z.
+        Used for periodic wrapping. Set to 0.0 to disable wrapping in z.
+
+    ntarget : int
+        Target number of voxels spanned by each kernel at its assigned level.
+        Controls the accuracy/speed trade-off. Default is 6.
```
Comment (kyleaoman) [inline swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py:1197]:
```suggestion
        Controls the accuracy/speed trade-off. Default is ``6``.
```

File: swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py [hunk 29]
```diff
@@ -0,0 +1,1225 @@
...
+
+    nlevels : int
+        Maximum number of coarsening levels in the hierarchy. Default is 4.
```
Comment (kyleaoman) [inline swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py:1200]:
```suggestion
        Maximum number of coarsening levels in the hierarchy. Default is ``4``.
```

File: swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py [hunk 30]
```diff
@@ -0,0 +1,1225 @@
...
+
+    Returns
+    -------
+    np.ndarray[np.float32, np.float32, np.float32]
```
Comment (kyleaoman) [inline swiftsimio/visualisation/volume_render_backends/nested_grids_scatter.py:1204]: Again hint on the dtype and shape is incorrectly formatted?

**Review by kyleaoman — CHANGES_REQUESTED** [review]
Hi Owen,

Sorry to leave this hanging for a while, for some reason I thought that you were still planning edits. Anyway some suggestions for you here, all fairly straightforward. Will want to make sure Josh is happy to merge this before we do, too.

**Comment by ojessop01** [comment]
Added 2d projections as well
